rafttest/interaction_env_handler_send_snapshot.go

```diff
@@ -0,0 +1,50 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rafttest
+
+import (
+	"testing"
+
+	"github.com/cockroachdb/datadriven"
+	"github.com/stretchr/testify/require"
+
+	"go.etcd.io/raft/v3"
+	"go.etcd.io/raft/v3/raftpb"
+)
+
+func (env *InteractionEnv) handleSendSnapshot(t *testing.T, d datadriven.TestData) error {
+	idxs := nodeIdxs(t, d)
+	require.Len(t, idxs, 2)
+	return env.SendSnapshot(idxs[0], idxs[1])
+}
+
+// SendSnapshot sends a snapshot.
+func (env *InteractionEnv) SendSnapshot(fromIdx, toIdx int) error {
+	snap, err := env.Nodes[fromIdx].Snapshot()
+	if err != nil {
+		return err
+	}
+	from, to := uint64(fromIdx+1), uint64(toIdx+1)
+	msg := raftpb.Message{
+		Type:     raftpb.MsgSnap,
+		Term:     env.Nodes[fromIdx].BasicStatus().Term,
+		From:     from,
+		To:       to,
+		Snapshot: &snap,
+	}
+	env.Messages = append(env.Messages, msg)
```

Comment:
>Do I understand correctly that we're modelling the case when snapshots are sent by things outside the raft RawNode's control (e.g. the CRDB's own snapshotting mechanism)?

That's right.

>Can we, though, trick the Node into sending this message instead of us making it up? I.e. is it possible that the node itself sends a snapshot, then somehow gets to sending a second snapshot at a higher index, and then ends up receiving a reply for the first snapshot but not the second?

I don't think we can. Once the node enters `StateSnapshot`, we can no longer increase the `PendingSnapshot` index because we return early here:

https://github.com/etcd-io/raft/blob/226ef1c2a6405b71c3538249f38c48ed0be169b7/raft.go#L620-L622

We could reset it by transitioning to `StateProbe` first, but afaict the only way to do that from `StateSnapshot` is to explicitly reject the snapshot with an `MsgSnapStatus`, and if we rejected the initial snapshot then we clearly did not apply it and we can't later tell the leader we did.

Besides, the case where snapshots are sent manually is the case we care about for our purposes, so it makes sense to test that.